[DOCS] Documentation Addition (Readme updates)

vidyasagar-amd opened the pull request:
## Proposed changes

This PR adds a series of missing READMEs and updates existing ones.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

vidyasagar-amd committed:
GH-2368 Adding a basic glossary

GH-2368 Minor edits

GH-2368 Adding missing READMEs and standardization.

resolving readme updates

GH-2368 Minor improvements to documentation.

Improving some readmes.

Further improvement for readmes.

Cleaned up the documentation in 'client_example' (#2468)

Update for PR

Update ACRONYMS.md to remove trivial terms

Update ACRONYMS.md to provide detailed explanations for BF16 and BF8 formats

Apply suggestion from @spolifroni-amd

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Apply suggestion from @spolifroni-amd

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Update README.md to clarify CK Tile API description and remove outdated references to the Tile Engine.

revise 37_transpose readme

revise 36_copy readme

Remove references to the Tile Engine in README files for 19_gemm_multi_d and 35_batched_transpose, and update distribution links for clarity.

Remove references to the Tile Engine in multiple README files and update distribution links for consistency and clarity.

Remove references to the Tile Engine in README files across multiple examples

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+# Acronyms in Composable Kernel
+
+This file lists technical acronyms found throughout the Composable Kernel codebase.
```
```suggestion
The following acronyms are used in the Composable Kernel codebase:
```

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| Acronym | Expansion | Explanation |
+|---------|-----------|-------------|
+| API     | Application Programming Interface | Interface for interacting with software components |
+| B2B     | Back-to-Back | Fused or sequential kernel operations (e.g., B2B GEMM) |
```
Does it described a fused op or a sequential op, or are fused and sequential ops the same thing?

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+|---------|-----------|-------------|
+| API     | Application Programming Interface | Interface for interacting with software components |
+| B2B     | Back-to-Back | Fused or sequential kernel operations (e.g., B2B GEMM) |
+| BF16    | Brain Floating Point 16 | 16-bit floating point format for ML workloads |
```
I would personally put in that it's equivalent to a float, but I don't know if that's more detail than you want here.

tenpercent replied:
"equivalent to float" is not precise. 1 sign bit, 8 exponent bits, 7 mantissa bits. E8M7 is another notation for that

AviralGoelAMD replied:
I changed the description, please have a relook @tenpercent

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| API     | Application Programming Interface | Interface for interacting with software components |
+| B2B     | Back-to-Back | Fused or sequential kernel operations (e.g., B2B GEMM) |
+| BF16    | Brain Floating Point 16 | 16-bit floating point format for ML workloads |
+| BF8     | 8-bit Brain Floating Point | 8-bit floating point format, variant for ML |
```
See BF16

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| BF16    | Brain Floating Point 16 | 16-bit floating point format for ML workloads |
+| BF8     | 8-bit Brain Floating Point | 8-bit floating point format, variant for ML |
+| BLAS    | Basic Linear Algebra Subprograms | Standardized library for linear algebra routines |
+| C++     | C Plus Plus | Programming language used for CK |
```
I probably wouldn't put this in :)

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| BF8     | 8-bit Brain Floating Point | 8-bit floating point format, variant for ML |
+| BLAS    | Basic Linear Algebra Subprograms | Standardized library for linear algebra routines |
+| C++     | C Plus Plus | Programming language used for CK |
+| CK      | Composable Kernel | The main project/library name |
```
```suggestion
| CK      | Composable Kernel | The shortened name of the Composable Kernel library. |
```

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| BLAS    | Basic Linear Algebra Subprograms | Standardized library for linear algebra routines |
+| C++     | C Plus Plus | Programming language used for CK |
+| CK      | Composable Kernel | The main project/library name |
+| CPU     | Central Processing Unit | General-purpose processor |
```
I'd mention that this is equivalent to HOST.

spolifroni-amd commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+| CPU     | Central Processing Unit | General-purpose processor |
+| CUDA    | Compute Unified Device Architecture | NVIDIA's GPU programming platform |
+| DLA     | Deep Learning Accelerator | Specialized hardware for deep learning workloads |
+| DRAM    | Dynamic Random-Access Memory | Main memory (global memory on GPU) |
```
```suggestion
| DRAM    | Dynamic Random-Access Memory | Main memory. Global memory on GPU |
```

AviralGoelAMD commented on ACRONYMS.md:
```diff
@@ -0,0 +1,81 @@
+
+| Acronym | Expansion | Explanation |
+|---------|-----------|-------------|
+| API     | Application Programming Interface | Interface for interacting with software components |
```
API seems too obvious.

AviralGoelAMD committed:
GH-2368 Adding a basic glossary

GH-2368 Minor edits

GH-2368 Adding missing READMEs and standardization.

resolving readme updates

GH-2368 Minor improvements to documentation.

Improving some readmes.

Further improvement for readmes.

Cleaned up the documentation in 'client_example' (#2468)

Update for PR

Update ACRONYMS.md to remove trivial terms

Update ACRONYMS.md to provide detailed explanations for BF16 and BF8 formats

Apply suggestion from @spolifroni-amd

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Apply suggestion from @spolifroni-amd

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

Update README.md to clarify CK Tile API description and remove outdated references to the Tile Engine.

revise 37_transpose readme

revise 36_copy readme

Remove references to the Tile Engine in README files for 19_gemm_multi_d and 35_batched_transpose, and update distribution links for clarity.

Remove references to the Tile Engine in multiple README files and update distribution links for consistency and clarity.

Remove references to the Tile Engine in README files across multiple examples

Refine README files by removing outdated references to the Tile Engine

tenpercent commented on ACRONYMS.md:
```diff
@@ -0,0 +1,68 @@
+|---------|-----------|-------------|
+| B2B     | Back-to-Back | Fused or sequential kernel operations (e.g., B2B GEMM) |
+| BF16    | Brain Floating Point 16 | 1 Signed bit, 8 Exponent bits, 7 Mantissa bits |
+| BF8     | 8-bit Brain Floating Point | 1 Signed bit, 3 Exponent bits, 4 Mantissa bits |
```
Thanks! Nit, 'Sign', not 'Signed'. Also checked wiki and IEEE calls 'Significand' what I was calling 'Mantissa' so that should be also changed

spolifroni-amd commented on client_example/01_gemm/README.md:
````diff
@@ -124,3 +141,40 @@ Table of supported cases by instance factory with XDL instruction for Row/Row/Ro
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

We need to avoid "execute" as much as possible

spolifroni-amd commented on client_example/02_gemm_add_add_fastgelu/README.md:
````diff
@@ -0,0 +1,67 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/03_gemm_layernorm/README.md:
````diff
@@ -0,0 +1,59 @@
+```bash
+cd composable_kernel/build
+make -j install
+```
````
See above re, the installation steps.

spolifroni-amd commented on client_example/03_gemm_layernorm/README.md:
````diff
@@ -0,0 +1,59 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/01_gemm/README.md:
```diff
@@ -124,3 +141,40 @@ Table of supported cases by instance factory with XDL instruction for Row/Row/Ro
+
+## How to Run
+
+### Prerequisites
```
```suggestion
### Prerequisites
```

Add the line:

"The Cmposable Kernel library must be installed"

I would actually point to the doc here instead of putting the make install instructions because the doc recommends using a docker container (and since I've tried building CK from scratch, this isn't a bad idea),

spolifroni-amd commented on client_example/02_gemm_add_add_fastgelu/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/04_contraction/README.md:
````diff
@@ -0,0 +1,58 @@
+cd composable_kernel/build
+make -j install
+```
+
````
See above re, the installation steps.

spolifroni-amd commented on client_example/04_contraction/README.md:
````diff
@@ -0,0 +1,58 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/05_layernorm/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/05_layernorm/README.md:
````diff
@@ -0,0 +1,68 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/06_softmax/README.md:
```diff
@@ -0,0 +1,56 @@
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/06_softmax/README.md:
````diff
@@ -0,0 +1,56 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/06_softmax/README.md:
```diff
@@ -0,0 +1,56 @@
+### Key Functions
+
+- **main()** (in `softmax4d.cpp`):  
+  Sets up input tensors, configures softmax parameters, launches the softmax kernel, and verifies the result.
```
Does Softmax need to be capitalized?

spolifroni-amd commented on client_example/07_grouped_convnd_fwd/README.md:
```diff
@@ -66,3 +79,54 @@ Table of supported cases by instance factory with fused elementwise operation:
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/07_grouped_convnd_fwd/README.md:
````diff
@@ -66,3 +79,54 @@ Table of supported cases by instance factory with fused elementwise operation:
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/08_fused_attention/README.md:
```diff
@@ -0,0 +1,91 @@
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/08_fused_attention/README.md:
````diff
@@ -0,0 +1,91 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/09_quantization/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/09_quantization/README.md:
````diff
@@ -0,0 +1,87 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/10_grouped_convnd_bwd_data/README.md:
```diff
@@ -46,3 +46,58 @@ Table of supported cases by instance factory with fused elementwise operation:
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/10_grouped_convnd_bwd_data/README.md:
````diff
@@ -46,3 +46,58 @@ Table of supported cases by instance factory with fused elementwise operation:
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/11_grouped_conv_bwd_weight/README.md:
```diff
@@ -60,3 +60,65 @@ Table of supported cases by instance factory with fused elementwise operation:
+
+## How to Run
+
+### Prerequisites
```
See above re, the installation steps.

spolifroni-amd commented on client_example/11_grouped_conv_bwd_weight/README.md:
````diff
@@ -60,3 +60,65 @@ Table of supported cases by instance factory with fused elementwise operation:
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/12_elementwise_normalization/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/12_elementwise_normalization/README.md:
````diff
@@ -0,0 +1,71 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/13_batchnorm/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/13_batchnorm/README.md:
````diff
@@ -0,0 +1,78 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/14_instance_id/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/14_instance_id/README.md:
````diff
@@ -0,0 +1,65 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/15_convnd_bwd_data/README.md:
```diff
@@ -0,0 +1,75 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/15_convnd_bwd_data/README.md:
````diff
@@ -0,0 +1,75 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/16_convnd_fwd/README.md:
```diff
@@ -0,0 +1,79 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/16_convnd_fwd/README.md:
````diff
@@ -0,0 +1,79 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/16_convnd_fwd/README.md:
```diff
@@ -0,0 +1,79 @@
+## Additional Details
+
+- Supports FP16, FP32, and FP8 composite for 3D convolution.
+- Example parameters can be adjusted in the source for different workloads.
```
I've seen this the other examples and I've been thinking that maybe it would be helpful to say _which_ parameters can be adjusted. I don't know. It's up to you.

I'm also thinking that you can just say "parameters" instead of "example parameters", since it's kind of obvious that these are the parameters that belong to the example. Unless it isn't and there are other parameters. In which case it would make sense to list which parameters can be modified.

I'm assuming there's also comments in the source code that indicates which parameters can be modified.

spolifroni-amd commented on client_example/17_grouped_gemm_fastgelu/README.md:
```diff
@@ -0,0 +1,73 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/17_grouped_gemm_fastgelu/README.md:
````diff
@@ -0,0 +1,73 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/18_groupnorm/README.md:
```diff
@@ -0,0 +1,82 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/18_groupnorm/README.md:
````diff
@@ -0,0 +1,82 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/19_pool/README.md:
```diff
@@ -0,0 +1,82 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/19_pool/README.md:
````diff
@@ -0,0 +1,82 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/20_splitk_gemm/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/20_splitk_gemm/README.md:
````diff
@@ -0,0 +1,68 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/21_grouped_gemm_bias/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/21_grouped_gemm_bias/README.md:
````diff
@@ -0,0 +1,67 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/22_grouped_gemm/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/22_grouped_gemm/README.md:
````diff
@@ -0,0 +1,78 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/23_elementwise_transpose/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/23_elementwise_transpose/README.md:
````diff
@@ -0,0 +1,66 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/24_grouped_conv_activation/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/24_grouped_conv_activation/README.md:
````diff
@@ -0,0 +1,90 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/25_wrapper/README.md:
```diff
@@ -1,13 +1,66 @@
+# Composable Kernel Wrapper GEMM Tutorial
 
-To implement these optimizations, you can use the CK wrapper or directly use available instances in CK. You can also refer to the [optimized GEMM example](https://github.com/ROCm/composable_kernel/blob/develop/client_example/25_wrapper/wrapper_optimized_gemm.cpp), that uses CK wrapper based on the [`gridwise_gemm_xdlops_v2r3`](https://github.com/ROCm/composable_kernel/blob/develop/include/ck/tens ...
+This tutorial demonstrates how to implement matrix multiplication (GEMM) using the Composable Kernel (CK) wrapper. The examples show both basic and optimized GEMM implementations, as well as how to use the wrapper for tensor transformations such as im2col.
```
Is this supposed to be "the examples in this tutorial"? Because it's one tutorial but it sounds like there are multiple examples.

spolifroni-amd commented on client_example/25_wrapper/README.md:
```diff
@@ -1,13 +1,66 @@
+---
+
+## How to Run
+
```
Is there no prerequisite for this one on purpose or was it an oversight?

spolifroni-amd commented on client_example/25_wrapper/README.md:
```diff
@@ -1,13 +1,66 @@
+
+## How to Run
+
+### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/25_wrapper/README.md:
```diff
@@ -168,5 +221,13 @@ The end result from `c_vgpr_reg` is stored in the `C` local partition (tensor pe
 
-If you want to dive deep into the details, you can find the entire example
-[here](https://github.com/ROCm/composable_kernel/blob/develop/client_example/25_wrapper/wrapper_basic_gemm.cpp).
+---
```
The other readmes don't have templates in them. Is there a reason why this one has it or was it an oversight that it was included?

vidyasagar-amd replied:
I think its there to explain how the wrapper functionality works

spolifroni-amd commented on client_example/26_reduce/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/26_reduce/README.md:
````diff
@@ -0,0 +1,66 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/27_im2col_col2im/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## Theory
+
+This client example demonstrates **im2col (image-to-column) and col2im (column-to-image) transformations**. These operations are used to convert image data into a matrix form suitable for GEMM-based convolution, and to reconstruct images from column representations.
```
```suggestion
This client example demonstrates **im2col (image-to-column) and col2im (column-to-image) transformations**. These operations are used to convert image data into a matrix form suitable for GEMM-based convolution and reconstruct images from column representations.
```

spolifroni-amd commented on client_example/27_im2col_col2im/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/27_im2col_col2im/README.md:
````diff
@@ -0,0 +1,70 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/28_gemm_mx/README.md:
```diff
@@ -0,0 +1,30 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/28_gemm_mx/README.md:
````diff
@@ -0,0 +1,30 @@
+make install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/29_gemm_add_multiply/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## Theory
+
+This client example demonstrates **GEMM fused with addition and multiplication operations**. This pattern is used in neural networks for bias addition, scaling, gating, and other elementwise transformations after a linear layer.
```
Is it bias addition, bias scaling, bias gating, or is bias addition, as well as scaling, gating?

vidyasagar-amd replied:
no its just 'bias addition', scaling and gating are different ops

spolifroni-amd commented on client_example/29_gemm_add_multiply/README.md:
```diff
@@ -0,0 +1,68 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/29_gemm_add_multiply/README.md:
````diff
@@ -0,0 +1,68 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/30_gemm_bf16Aint8B/README.md:
```diff
@@ -0,0 +1,89 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/30_gemm_bf16Aint8B/README.md:
````diff
@@ -0,0 +1,89 @@
+make install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on client_example/31_grouped_gemm_bf16Aint8B/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on client_example/31_grouped_gemm_bf16Aint8B/README.md:
````diff
@@ -0,0 +1,90 @@
+make install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/01_gemm/README.md:
```diff
@@ -1,27 +1,224 @@
+
+## Introduction
+
+GEMM (General Matrix Multiplication) is a fundamental operation in linear algebra and deep learning. It computes the product of two matrices, optionally adds a bias or residual, and is the core of many neural network layers (MLPs, attention, convolutions via im2col). This example demonstrates the flexible and high-performance GEMM API provided by Composable Kernel (CK).
```
In most of the readmes -- I haven't checked all of them -- Composable Kernel is referred to by its full name and not by CK. Here CK is in parentheses, but it doesn't appear in any of the readmes I've seen so far.

Do you want to add it everywhere, or removed it from here? You can also leave this as-is, but I think consistency is nice.

spolifroni-amd commented on example/01_gemm/README.md:
```diff
@@ -1,27 +1,224 @@
+- $C$: [M, N] output
+- $\alpha, \beta$: scalars (often 1.0, 0.0)
+
+GEMM is implemented using a tiled/blocking strategy to maximize data reuse and memory bandwidth. Modern GPU implementations use matrix core/XDL/MFMA instructions for high throughput. The operation is the computational backbone for transformer attention, MLPs, CNNs (via lowering), and more.
```
So the term "modern GPUs" feels a bit strange, and this theory section seems to deviate from the other theory sections. It feels more marketing-ish, like it's trying to sell me on the merits of GEMMs. But we already know the merits of GEMMs. I would maybe streamline this.

spolifroni-amd commented on example/01_gemm/README.md:
```diff
@@ -1,27 +1,224 @@
+## CK GEMM API Overview
+
+CK provides a highly composable GEMM API via the `DeviceGemm` family of device operations. These are highly templated to support a wide range of data types, layouts, and fused operations.
+
```
Again this section -- actually this entire readme -- seems to deviate from the rest. I don't know if that's intentional, but it feels more like something that would be in the documentation than in a readme attached to an example. 

If you want to add it to the doc and then streamline the readme, let me know.

vidyasagar-amd replied:
yes, it would be great to pull this information into a tutorial, thats in the works

spolifroni-amd commented on example/01_gemm/README.md:
```diff
@@ -1,27 +1,224 @@
+- **DeviceGemmStreamK**: Stream K using AtomicAdd
+
+---
+
```
The prerequisites seems to be missing from here.

spolifroni-amd commented on example/01_gemm/README.md:
```diff
@@ -1,27 +1,224 @@
+
+## How to Run
+
+### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on example/01_gemm/README.md:
````diff
@@ -1,27 +1,224 @@
+    └── reference_gemm.hpp    # CPU reference GEMM for correctness checking
 ```
+
+### Key Classes and Functions
````
Again this is a major deviation and I don't know if you want code in the readme.

spolifroni-amd commented on example/02_gemm_bilinear/README.md:
```diff
@@ -1,6 +1,74 @@
+## Introduction
+
+This example demonstrates GEMM (General Matrix Multiplication) fused with bilinear operations on auxiliary tensors using Composable Kernel (CK). Bilinear fusion patterns are widely used in neural networks for gating, attention, and multimodal feature fusion, where the output of a matrix multiplication is combined elementwise with one or more additional tensors.
+
```
Going to point out the CK/Composable Kernel duality.

spolifroni-amd commented on example/02_gemm_bilinear/README.md:
```diff
@@ -1,6 +1,74 @@
+- **DeviceGemmMultipleD_Bilinear**: GEMM with bilinear fusion in the epilogue
+
+---
+
```
seems to be missing the prerequisites.

spolifroni-amd commented on example/02_gemm_bilinear/README.md:
```diff
@@ -1,6 +1,74 @@
+
+## How to Run
+
+### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on example/02_gemm_bilinear/README.md:
````diff
@@ -9,3 +77,42 @@
+    └── element_wise_operation.hpp       # Elementwise operation definitions
+```
+
+### Key Classes and Functions
````
Back to the question of code vs no code.

spolifroni-amd commented on example/03_gemm_bias_relu/README.md:
```diff
@@ -1,10 +1,66 @@
+**Algorithmic Background:**
+- The GEMM result is kept in registers, bias is added, and ReLU is applied before writing to global memory.
+- This fusion eliminates intermediate memory traffic and is a standard optimization in deep learning frameworks.
+
```
prerequisites seem to be missing.

spolifroni-amd commented on example/03_gemm_bias_relu/README.md:
```diff
@@ -1,10 +1,66 @@
+
+## How to Run
+
+### Build and Execute
```
```suggestion
### Build and run
```

spolifroni-amd commented on example/03_gemm_bias_relu/README.md:
````diff
@@ -1,10 +1,66 @@
+    └── element_wise_operation.hpp         # Elementwise operation definitions
+```
+
+### Key Classes and Functions
````
The code v no-code question

spolifroni-amd commented on example/04_gemm_add_add_fastgelu/README.md:
```diff
@@ -1,10 +1,79 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/04_gemm_add_add_fastgelu/README.md:
````diff
@@ -1,10 +1,79 @@
+make -j install
 ```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/04_gemm_add_add_fastgelu/README.md:
````diff
@@ -1,10 +1,79 @@
+    └── element_wise_operation.hpp         # Elementwise operation definitions
+```
+
+### Key Classes and Functions
````
Code vs no-code

spolifroni-amd commented on example/09_convnd_fwd/README.md:
```diff
@@ -1,6 +1,44 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/09_convnd_fwd/README.md:
````diff
@@ -1,6 +1,44 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/09_convnd_fwd/README.md:
````diff
@@ -16,3 +54,36 @@
+    └── blockwise_convnd_fwd_xdl.hpp # Block-level convolution
+```
+
+### Key Classes and Functions
````
code vs no-code

spolifroni-amd commented on example/10_convnd_fwd_multiple_d_multiple_reduce/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/10_convnd_fwd_multiple_d_multiple_reduce/README.md:
````diff
@@ -0,0 +1,67 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/10_convnd_fwd_multiple_d_multiple_reduce/README.md:
````diff
@@ -0,0 +1,67 @@
+    └── gridwise_convnd_fwd_multiple_d_multiple_reduce.hpp # Grid-level kernel
+```
+
+### Key Classes and Functions
````
code vs no-code

spolifroni-amd commented on example/11_convnd_fwd_bias/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/11_convnd_fwd_bias/README.md:
````diff
@@ -0,0 +1,67 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/11_convnd_fwd_bias/README.md:
````diff
@@ -0,0 +1,67 @@
+    └── gridwise_convnd_fwd_bias.hpp     # Grid-level kernel
+```
+
+### Key Classes and Functions
````
code vs no code

spolifroni-amd commented on example/12_reduce/README.md:
```diff
@@ -1,6 +1,40 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/12_reduce/README.md:
````diff
@@ -1,6 +1,40 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/12_reduce/README.md:
````diff
@@ -60,3 +98,36 @@ Warm up 1 time
+    └── blockwise_reduce.hpp    # Block-level reduction
+```
+
+### Key Classes and Functions
````
code vs no code

spolifroni-amd commented on example/13_pool2d_fwd/README.md:
```diff
@@ -1,6 +1,43 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/13_pool2d_fwd/README.md:
````diff
@@ -1,6 +1,43 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/13_pool2d_fwd/README.md:
````diff
@@ -39,3 +78,37 @@ Warm up 1 time
+    └── blockwise_pool.hpp        # Block-level pooling
+```
+
+### Key Classes and Functions
````
code vs no code

spolifroni-amd commented on example/14_gemm_quantization/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/14_gemm_quantization/README.md:
````diff
@@ -0,0 +1,70 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/14_gemm_quantization/README.md:
````diff
@@ -0,0 +1,70 @@
+```
+
+### Key Classes and Functions
+
````
code vs no code

spolifroni-amd commented on example/15_grouped_gemm/README.md:
```diff
@@ -1,9 +1,73 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/15_grouped_gemm/README.md:
````diff
@@ -1,9 +1,73 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/15_grouped_gemm/README.md:
````diff
@@ -1,9 +1,73 @@
+│   └── gridwise_grouped_gemm_xdl.hpp     # Grid-level grouped GEMM kernel
+```
+
+### Key Classes and Functions
````
code vs no code

spolifroni-amd commented on example/16_gemm_multi_d_multi_reduces/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## How to Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/16_gemm_multi_d_multi_reduces/README.md:
````diff
@@ -0,0 +1,66 @@
+make -j install
+```
+
+### Build and Execute
````
```suggestion
### Build and run
```

spolifroni-amd commented on example/16_gemm_multi_d_multi_reduces/README.md:
````diff
@@ -0,0 +1,66 @@
+```
+
+### Key Classes and Functions
+
````
code vs no code

spolifroni-amd commented on example/17_convnd_bwd_data/README.md:
```diff
@@ -1,6 +1,65 @@
+
+## Build and Run
+
+### Prerequisites
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/18_batched_gemm_reduce/README.md:
```diff
@@ -0,0 +1,81 @@
+
+1.  **Batch Scheduling**: The `BatchCount` GEMM problems are distributed across the GPU's thread blocks. Each block is assigned one or more GEMMs from the batch to compute.
+
+2.  **Tiled GEMM Core**: For each assigned GEMM, the thread block executes a standard tiled GEMM algorithm to compute the product $A_{[b]} \times B_{[b]}$. The result for each tile of $C_{[b]}$ is accumulated in the private registers of the threads.
```
```suggestion
2.  **Tiled GEMM Core**: For each assigned GEMM, the thread block runs a standard tiled GEMM algorithm to compute the product $A_{[b]} \times B_{[b]}$. The result for each tile of $C_{[b]}$ is accumulated in the private registers of the threads.
```

spolifroni-amd commented on example/18_batched_gemm_reduce/README.md:
```diff
@@ -0,0 +1,81 @@
+## Build and Run
+
+### Prerequisites
+Ensure the Composable Kernel library is built and installed.
```
See the very first comment about how to approach this.

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+1.  **Simple Element-wise**: Applying a binary function to two input tensors of the *same* shape.
+2.  **Element-wise with Broadcasting**: Applying a binary function to two input tensors of *different but compatible* shapes.
+
+Broadcasting defines a set of rules for applying element-wise operations on tensors of different sizes, and it is a cornerstone of libraries like NumPy and TensorFlow.
```
So here the term broadcasting is introduced, but it's sort of written backwards. Like I'd say "a set of rules for applying elementwise operations on tensors. [...] is referred to as broadcasting"

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+$C_{i,j,k,\dots} = A_{i,j,k,\dots} \odot B_{i,j,k,\dots}$
+
+### Element-wise with Broadcasting
+Broadcasting allows element-wise operations on tensors with different shapes, provided they are compatible. Two dimensions are compatible if they are equal, or if one of them is 1. The operation implicitly "stretches" or "duplicates" the tensor with the dimension of size 1 to match the other tensor's shape.
```
Since broadcating was already defined, we can skip redefining it here.

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+
+$C_{i,j,k,\dots} = A_{i,j,k,\dots} \odot B_{i,j,k,\dots}$
+
+### Element-wise with Broadcasting
```
```suggestion
### Elementwise with Broadcasting
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+# Binary Element-wise Operations with Broadcasting
```
```suggestion
# Binary Elementwise Operations with Broadcasting
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+# Binary Element-wise Operations with Broadcasting
+
+This example demonstrates a generic binary element-wise operation, a fundamental building block in numerical computing. It covers two important cases:
```
```suggestion
This example demonstrates a generic binary elementwise operation, a fundamental building block in numerical computing. It covers two important cases:
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+# Binary Element-wise Operations with Broadcasting
+
+This example demonstrates a generic binary element-wise operation, a fundamental building block in numerical computing. It covers two important cases:
+1.  **Simple Element-wise**: Applying a binary function to two input tensors of the *same* shape.
```
```suggestion
1.  **Simple Elementwise**: Applying a binary function to two input tensors of the *same* shape.
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+
+This example demonstrates a generic binary element-wise operation, a fundamental building block in numerical computing. It covers two important cases:
+1.  **Simple Element-wise**: Applying a binary function to two input tensors of the *same* shape.
+2.  **Element-wise with Broadcasting**: Applying a binary function to two input tensors of *different but compatible* shapes.
```
```suggestion
2.  **Elementwise with Broadcasting**: Applying a binary function to two input tensors of *different but compatible* shapes.
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+
+Here, the single row of `B` is broadcast across all `M` rows of `A`. The output tensor `C` has the shape `(M, N)`.
+
+Common binary element-wise operations include addition, subtraction, multiplication (Hadamard product), division, max, and min.
```
```suggestion
Common binary elementwise operations include addition, subtraction, multiplication (Hadamard product), division, max, and min.
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
````diff
@@ -0,0 +1,102 @@
+
+        C[out_idx] = A[a_idx] op B[b_idx];
+    }
+    ```
````
do we want code in the readme?

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+-   The core of the broadcasting logic lies in the `get_broadcast_coord` function. If an input tensor's dimension is 1, the coordinate for that dimension is always set to 0, effectively reusing the same element across the broadcast dimension. If the dimension matches the output, the coordinate is passed through.
+-   This strategy ensures that memory accesses to the larger tensor remain coalesced, while accesses to the smaller, broadcasted tensor will naturally involve re-reading the same values, which is efficiently handled by the GPU's cache hierarchy.
+
+Like the simple case, broadcasted element-wise operations are almost always memory-bandwidth-bound.
```
```suggestion
Like the simple case, broadcasted elementwise operations are almost always memory-bandwidth-bound.
```

spolifroni-amd commented on example/19_binary_elementwise/README.md:
```diff
@@ -0,0 +1,102 @@
+## Build and Run
+
+### Prerequisites
+Ensure the Composable Kernel library is built and installed.
```
See the first comment about how to handle this.

vidyasagar-amd committed:
Updates based on PR feedback 1

vidyasagar-amd committed:
Updates based on PR feedback 2

vidyasagar-amd committed:
Updates based on PR feedback 3

vidyasagar-amd committed:
Updates based on PR feedback 4

vidyasagar-amd committed:
Updates based on PR feedback 5

vidyasagar-amd committed:
Updates based on PR feedback 6

vidyasagar-amd committed:
Updates based on PR feedback 7

vidyasagar-amd committed:
Updates based on PR feedback 8

ThomasNing committed:
Content Modification of CK Tile Example

ThomasNing committed:
Modify the ck_tile gemm config

shumway commented on example/30_grouped_conv_fwd_multiple_d/README.md:
```diff
@@ -1,4 +1,63 @@
+1.  **Group Scheduling**: The `G` independent problems are distributed across the GPU's thread blocks. Each thread block is assigned to compute the fused convolution for one of the `G` groups.
+
+2.  **Implicit GEMM Core**: Once a thread block is assigned a group `g`, it executes the convolution for that group using the implicit GEMM algorithm. This involves:
+    -   Calculating the base memory addresses for the group's input tensors: $\text{In}_{[g]}, \text{W}_{[g]}, D_{0[g]}, \dots, E_{[g]}$.
```
For some reason this formatted math isn't rendering in markdown. I can't see any mistakes. It may be a bug in the GitHub markdown formatter.

vidyasagar-amd replied:
strange, it renders for me in VSCode but must be something with the github formatter
<img width="983" height="626" alt="image" src="https://github.com/user-attachments/assets/3bd422d3-d090-41c5-b3df-ae2e2ea3202b" />
